### ISSUE TITLE
Make sure that get_file_type is not called until configurationDone is called.

### ISSUE THREAD

**Issue by fabioz** [Microsoft/ptvsd#1355]
- [ ] Set ignore patterns in handle launch/attach (see comment in wrapper about this issue)
- [ ] Forward setDebuggerProperty request to pydevd
- [ ] Remove `dont_trace_external_files` usage from wrapper.
- [ ] Ensure settings set setDebuggerProperty are applied

**Comment by fabioz** [comment]
It seems that `ptvsd.attach_server.break_into_debugger` calls `get_file_type` to discover which frame it should stop into.

**Comment by karthiknadig** [comment]
Break into debugger should be blocked until configurationDone is called on ptvsd. You are right this is another integration point that will need custom protocol support.

**Comment by fabioz** [comment]
The frame is used as the last thing when attaching, so, I think that accepting a callable for the frame that's computed on demand may be enough here (I'm checking that now).

**Comment by fabioz** [comment]
@karthiknadig I'm actually having many other failures after I integrated my latest changes into the master (because `_forward_request_to_pydevd` is now sending `success=False` on `pydevd` failures which it wasn't doing properly before).

So, I'm still investigating as it's not only `test_breakpoint_function` which seems to be breaking now (so, the change I proposed is not enough -- still investigating).

**Comment by fabioz** [comment]
@karthiknadig investigating more it seems that besides the problem with `get_file_type` being called on `ptvsd` on `break_into_debugger` (for which I've done a fix at: https://github.com/fabioz/ptvsd/commit/2fb2f574f1734441c52f10d033edb531ad88dedd),  `PyDB.ready_to_run` (which is the flag in `pydevd` which says that everything is setup) is being set to True before `configurationDone`.

`get_file_type` is being called during the `setup_connection` (see the stack below) after the `ready_to_run` flag is set to True but before `on_setdebuggerproperty_request` is called -- it happens reliably on tests for me (when `start_method` == `attach_socket_cmdline`).

```
  File "/home/fabioz/miniconda3/envs/py37/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/fabioz/miniconda3/envs/py37/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/fabioz/Desktop/ptvsd/ptvsd/src/ptvsd/__main__.py", line 405, in <module>
    main(sys.argv)
  File "/home/fabioz/Desktop/ptvsd/ptvsd/src/ptvsd/__main__.py", line 391, in main
    run()
  File "/home/fabioz/Desktop/ptvsd/ptvsd/src/ptvsd/__main__.py", line 256, in run_file
    setup_connection()
  File "/home/fabioz/Desktop/ptvsd/ptvsd/src/ptvsd/__main__.py", line 249, in setup_connection
    daemon = ptvsd._remote.enable_attach(addr)
  File "/home/fabioz/Desktop/ptvsd/ptvsd/src/ptvsd/_remote.py", line 97, in enable_attach
    patch_multiprocessing=ptvsd.options.multiprocess)
  File "/home/fabioz/Desktop/ptvsd/ptvsd/src/ptvsd/_vendored/pydevd/pydevd.py", line 1857, in settrace
    stop_at_frame,
  File "/home/fabioz/Desktop/ptvsd/ptvsd/src/ptvsd/_vendored/pydevd/pydevd.py", line 1929, in _locked_settrace
    debugger.set_trace_for_frame_and_parents(get_frame().f_back)
  File "/home/fabioz/Desktop/ptvsd/ptvsd/src/ptvsd/_vendored/pydevd/pydevd.py", line 1474, in set_trace_for_frame_and_parents
    file_type = self.get_file_type(abs_path_real_path_and_base)
  File "/home/fabioz/Desktop/ptvsd/ptvsd/src/ptvsd/_vendored/pydevd/pydevd.py", line 596, in get_file_type
```

So, to actually fix this the way that the connection happens needs some tweaks to only set `PyDB.ready_to_run` to True after `configurationDone` -- or at least after the `on_setdebuggerproperty_request` (and the fix to `break_into_debugger` I've done in https://github.com/fabioz/ptvsd/commit/2fb2f574f1734441c52f10d033edb531ad88dedd also needs to be applied)... I'm actually surprised we don't have other problems because of that as setting `ready_to_run` before `configurationDone` is called could be a cause for other racing conditions.

@karthiknadig I think @int19h worked more on the connection setup code, so, he may be better suited to handle this (so, unassigning myself for now).

**Comment by int19h** [comment]
The reason why we're setting `ready_to_run` early is because it's currently the only way to unblock `pydevd.settrace()`, which is necessary for `enable_attach()` to return immediately rather than block waiting for connection. In tests, it doesn't make a difference because we always do `enable_attach` followed by `wait_for_attach` there, so if you move that assignment, it behaves the same. But there are scenarios where `wait_for_attach` is not used, typically with web apps.

So, basically, this needs #1248. Once we have that, it's safe to completely get rid of `ready_to_run` in ptvsd - we already issue `CMD_RUN` in `configurationDone`, which sets it - which will fix this issue.

**Comment by fabioz** [comment]
Investigating, it seems it's reasonable to clear the caches which would be affected (so, `get_file_type` can be called at any time).

From what I could see, the caches are:

```
_CACHE_FILE_TYPE (needs api to clear)


_clear_skip_caches() will clear:
-> global_cache_skips
-> global_cache_frame_skips

_clear_filters_caches() will clear:
-> _in_project_scope_cache
-> _exclude_by_filter_cache
    
    
Change PyDB.mtime 
-> FuncCodeInfo.always_skip_code (note: need to set tracing in existing frames in this case).
```

I'll go on and make this change. Note that I feel https://github.com/Microsoft/ptvsd/issues/1248 is a different item (will add some more comments there on some things which I think could go awry regardless of this issue).
